typing/typecore.ml

```diff
@@ -4251,14 +4247,15 @@ and type_expect_
       if label.lbl_mut = Immutable then
         raise(Error(loc, env, Label_not_mutable lid.txt));
       rue {
-        exp_desc = Texp_setfield(record, label_loc, label, newval);
+        exp_desc = Texp_setfield(record,
+          Value_mode.regional_to_global_alloc rmode,
```

Comment:
I think this is a bug. The mode on `Texp_setfield` is not really an allocation mode, in that it does not describe where something should be allocated. Instead it describes whether the record being written to might be on the stack. If it is on the stack then we need to do a slightly different write barrier. In particular, that means that a regional record must be marked as "possibly on the stack", whilst this code is marking it "definitely on the heap".

It would probably be a good idea to just use a different type for this data to try and avoid making this mistake again in future. (For the third time, since this was also a bug in earlier versions of the compiler).

Reply:
Indeed it is a bug - originally it was `regional_to_local_alloc`. Yes I will define another type, named either "value_mode" or "access_mode" - which one is better?

Reply:
Maybe "write_barrier_mode" or "modify_mode"?

Reply:
Do you mean to have a new type to replace `alloc_mode` in `Texp_setfield`, or in `Psetfield`? I.e. what stage should this new type be added to?